src/jules_pytk/config/inputs.py

```diff
@@ -0,0 +1,117 @@
+from os import PathLike
+from pathlib import Path
+from typing import Any
+
+import numpy
+import xarray
+
+from jules_pytk.config.namelists import JulesNamelists
+
+__all__ = [
+    "read_ascii",
+    "write_ascii",
+    "read_netcdf",
+    "write_netcdf",
+    "JulesInput",
+]
+
+def read_ascii(path: str | PathLike) -> numpy.ndarray:
+    with open(path, "r") as file:
+
+        # TODO: comment not being used
+        first_line = file.readline().strip()
+        if first_line[0] in ("#", "!"):  # Treated as comments by JULES
+            comment = first_line
+        else:
+            comment = ""
+
+    data = numpy.loadtxt(path, comments=("#", "!"))
+
+    return data
+
+
+def write_ascii(data: numpy.ndarray, path: str | PathLike, comment="") -> None:
+    numpy.savetxt(path, data, header=comment)
+
+
+def read_netcdf(path: str | PathLike, lazy: bool = True) -> xarray.Dataset:
+    # NOTE: loads entire dataset into memory. Bad idea for big files
+    # TODO: lazy
+    return xarray.load_dataset(path)
+
+
+def write_netcdf(data: xarray.Dataset, path: str | PathLike) -> None:
+    data.to_netcdf(path)
+
+
+class JulesInput:
+    def __init__(self, path_in_namelist: str | PathLike) -> None:
+        path_in_namelist = Path(path_in_namelist)
+
+        match Path(path_in_namelist).suffix:
+            case ".nc" | ".cdf":
+                self._loader = read_netcdf
+                self._dumper = write_netcdf
+            case ".asc" | ".txt" | ".dat":
+                self._loader = read_ascii
+                self._dumper = write_ascii
+            case _:
+                raise ValueError(f"Invalid suffix: {path_in_namelist.suffix}")
+
+        self._path = path_in_namelist
+        self._data = None
```

Comment:
The basic idea with this class is that the first-class citizen is the path appearing in the namelists - not any particular file on the disk, or any data.

However, the class extends this simple path by providing mechanisms for associating data with it - essentially by doing `input.data = data`.

The assumption is that by the time you want to dump your configuration and run `jules.exe`, you have associated some data (in memory) with each _relative_ path in the namelist. This data is then written along with the namelists.